### PR TITLE
elf: add Module.LookupNextElement

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

**Comment by safchain** [comment]
Hi, any news on this ? thanks

File: elf/table.go [hunk 1]
```diff
@@ -131,3 +140,33 @@ func (b *Module) DeleteElement(mp *Map, key unsafe.Pointer) error {
 
 	return nil
 }
+
+// LookupNextElement lookups the next element in the mp.
+// The key and the value are stored in the key and value unsafe.Pointer.
```
Comment (schu) [inline elf/table.go]: Should this be "The key and the value are stored in the nextKey and value unsafe.Pointer."?

Comment (safchain) [reply elf/table.go]: @schu Thanks for reviewing. You're right "nextKey". I will try to introduce tests for that. I'll try to add test for other operations too as it seems there is currently no test.

File: bpf_test.go [hunk 2]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+			t.Fatalf("key %d found", nextKey)
+		}
+		expected[nextKey] = true
+
+		key = nextKey
+	}
+
+	for key, found := range expected {
+		if !found {
+			t.Fatalf("expeted key %d not found", key)
```
Comment (iaguis) [inline bpf_test.go]: expeted -> expected

File: elf/table.go [hunk 3]
```diff
@@ -131,3 +140,33 @@ func (b *Module) DeleteElement(mp *Map, key unsafe.Pointer) error {
 
 	return nil
 }
+
+// LookupNextElement lookups the next element in the mp using the given key.
```
Comment (iaguis) [inline elf/table.go]: lookups -> looks up
in the mp -> in mp

File: elf/table.go [hunk 4]
```diff
@@ -131,3 +140,33 @@ func (b *Module) DeleteElement(mp *Map, key unsafe.Pointer) error {
...
+// The next key and the value are stored in the nextKey and value parameter.
+// Returns false at the end of the mp.
```
Comment (iaguis) [inline elf/table.go]: at the end of the mp -> at the end of mp

File: bpf_test.go [hunk 5]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+func checkUpdateDeleteElement(t *testing.T, b *elf.Module) {
+	mp := b.Map("dummy_hash")
+	if mp == nil {
+		t.Fatal("unable to find dummy_hash map")
+	}
+
+	key := 1000
+	value := 1000
+	if err := b.UpdateElement(mp, unsafe.Pointer(&key), unsafe.Pointer(&value), BPF_ANY); err != nil {
+		t.Fatal(err)
```
Comment (iaguis) [inline bpf_test.go]: Can you add some context to these errors? Something like:

```
t.Fatal("Failed trying to update an element with BPF_ANY")
```

And in the next one:

```
t.Fatal("Succeeded updating element with BPF_NOEXIST, but an element with the same key was added to the map before") 
```

and so on.

File: bpf_test.go [hunk 6]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+func checkLookupElement(t *testing.T, b *elf.Module) {
+	mp := b.Map("dummy_hash")
+	if mp == nil {
+		t.Fatal("unable to find dummy_hash map")
+	}
+
+	key := 2000
+	value := 2000
+	if err := b.UpdateElement(mp, unsafe.Pointer(&key), unsafe.Pointer(&value), BPF_ANY); err != nil {
+		t.Fatal(err)
```
Comment (iaguis) [inline bpf_test.go]: Same thing for these errors, if something fails we should be able to figure out what the problem was at a glance.

File: bpf_test.go [hunk 7]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+	if value != lvalue {
+		t.Fatalf("wrong value returned, expected %d, got %d", value, lvalue)
+	}
+
+	key = 3000
+	if err := b.LookupElement(mp, unsafe.Pointer(&key), unsafe.Pointer(&lvalue)); err == nil {
+		t.Fatal(err)
+	}
+
+	expected := map[int]bool{2000: false}
```
Comment (iaguis) [inline bpf_test.go]: This map is a bit confusing with that name. I'd call this `found` and the `found` below just `f`.

**Comment by safchain** [comment]
@schu @iaguis I addressed you comments, could you please have another look

File: bpf_test.go [hunk 8]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+	if err := b.UpdateElement(mp, unsafe.Pointer(&key), unsafe.Pointer(&value), BPF_EXIST); err != nil {
+		t.Fatal("failed trying to update an element with BPF_EXIST while the key was added to the map before")
+	}
+
+	if err := b.DeleteElement(mp, unsafe.Pointer(&key)); err != nil {
+		t.Fatal("failed to delete an element")
+	}
+
+	if err := b.UpdateElement(mp, unsafe.Pointer(&key), unsafe.Pointer(&value), BPF_EXIST); err == nil {
+		t.Fatal("succeeded updating element with BPF_EXIST, but the element was deleted to the map before")
```
Comment (schu) [inline bpf_test.go]: `s/deleted to/deleted from/`

File: bpf_test.go [hunk 9]
```diff
@@ -253,11 +261,105 @@ func checkPinConfigCleanup(t *testing.T, expectedPaths []string) {
...
+			t.Fatalf("key %d found", nextKey)
+		}
+		found[nextKey] = true
+
+		key = nextKey
+	}
+
+	for key, f := range found {
+		if !f {
+			t.Fatalf("expeted key %d not found", key)
```
Comment (schu) [inline bpf_test.go]: `s/expeted/expected/`

File: elf/table.go [hunk 10]
```diff
@@ -131,3 +140,33 @@ func (b *Module) DeleteElement(mp *Map, key unsafe.Pointer) error {
 
 	return nil
 }
+
+// LookupNextElement look up the next element in mp using the given key.
```
Comment (schu) [inline elf/table.go]: `s/look/looks/`

**Comment by safchain** [comment]
@schu just submitted a new version with typo fixes

**Comment by schu** [comment]
Thanks again! :)

**Pull request merged**